solr/solrj-streaming/src/java/org/apache/solr/client/solrj/io/stream/FeaturesSelectionStream.java

```diff
@@ -352,56 +354,49 @@ public Tuple read() throws IOException {
-            String term = shardTopTerms.getName(i);
-            double score = shardTopTerms.getVal(i);
-            int docFreq = shardDocFreqs.get(term);
-            double prevScore = termScores.containsKey(term) ? termScores.get(term) : 0;
-            long prevDocFreq = docFreqs.containsKey(term) ? docFreqs.get(term) : 0;
-            termScores.put(term, prevScore + score);
-            docFreqs.put(term, prevDocFreq + docFreq);
-          }
-        }
-
-        List<Tuple> tuples = new ArrayList<>(numTerms);
-        termScores = sortByValue(termScores);
-        int index = 0;
-        for (Map.Entry<String, Double> termScore : termScores.entrySet()) {
-          if (tuples.size() == numTerms) break;
-          index++;
-          Tuple tuple = new Tuple();
-          tuple.put(ID, featureSet + "_" + index);
-          tuple.put("index_i", index);
-          tuple.put("term_s", termScore.getKey());
-          tuple.put("score_f", termScore.getValue());
-          tuple.put("featureSet_s", featureSet);
-          long docFreq = docFreqs.get(termScore.getKey());
-          double d = Math.log(((double) numDocs / (double) (docFreq + 1)));
-          tuple.put("idf_d", d);
-          tuples.add(tuple);
+          shardTopTerms.forEach(
+              (term, score) -> {
+                int docFreq = shardDocFreqs.get(term);
+                termScores.merge(term, score, Double::sum);
+                docFreqs.merge(term, (long) docFreq, Long::sum);
+              });
         }
-
-        tuples.add(Tuple.EOF());
-
-        tupleIterator = tuples.iterator();
+        final long numDocsF = numDocs; // make final
+
+        final AtomicInteger idGen = new AtomicInteger(1);
+
+        tupleIterator =
+            termScores.entrySet().stream()
+                .sorted( // sort by score descending
+                    Comparator.<Map.Entry<String, Double>>comparingDouble(Entry::getValue)
+                        .reversed())
+                .limit(numTerms)
+                .map(
+                    (termScore) -> {
+                      int index = idGen.getAndIncrement();
+                      Tuple tuple = new Tuple();
+                      tuple.put(ID, featureSet + "_" + index);
+                      tuple.put("index_i", index);
+                      tuple.put("term_s", termScore.getKey());
+                      tuple.put("score_f", termScore.getValue());
+                      tuple.put("featureSet_s", featureSet);
+                      long docFreq = docFreqs.get(termScore.getKey());
+                      double d = Math.log(((double) numDocsF / (double) (docFreq + 1)));
+                      tuple.put("idf_d", d);
+                      return tuple;
+                    })
+                .iterator();
+      }
+      if (tupleIterator.hasNext()) {
+        return tupleIterator.next();
+      } else {
+        return Tuple.EOF();
       }
-
-      return tupleIterator.next();
     } catch (Exception e) {
```

Comment:
I didn't touch this but I'm very tempted.  This try-catch is sloppy for multiple reasons.  First, catching `Exception` in particular should be done sparingly.  This code wraps RuntimeExceptions and even existing IOExceptions as an IOException.  Surely this isn't right.  Furthermroe, this try-catch is so all-encompassing, including logic like the iterator stuff that shouldn't belong within it's super long scope.  It overall makes the method hard to read (cyclomatic complexity).